includes/Story_Renderer.php

```diff
@@ -0,0 +1,191 @@
+			$replacement = sprintf( '<html amp %s>', get_language_attributes( 'html' ) );
+			$result      = preg_replace( self::HTML_STRUCTURE_HTML_START_TAG, $replacement, $content, 1 );
+
+			if ( null !== $result ) {
+				return $result;
+			}
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Returns the full HTML <head> markup for a given story besides boilerplate.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function get_html_head_markup() {
+		ob_start();
+		?>
+		<meta name="generator" content="<?php printf( 'Web Stories %s', esc_attr( WEBSTORIES_VERSION ) ); ?>" />
+		<?php
+
+		/**
+		 * Prints scripts or data in the head tag on the front end.
+		 */
+		do_action( 'web_stories_story_head' );
+
+		return (string) ob_get_clean();
+	}
+
+	/**
+	 * Replaces markers in HTML <head> with dynamic content.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function replace_html_head( $content ) {
+		$start_tag = '<meta name="web-stories-replace-head-start"/>';
+		$end_tag   = '<meta name="web-stories-replace-head-end"/>';
+
+		$start_tag_pos = strpos( $content, $start_tag );
+		$end_tag_pos   = strpos( $content, $end_tag );
+
+		if ( false !== $start_tag_pos && false !== $end_tag_pos ) {
+			$end_tag_pos += strlen( $end_tag );
+			$content      = substr_replace( $content, $this->get_html_head_markup(), $start_tag_pos, $end_tag_pos - $start_tag_pos );
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Adds square, and landscape poster images to the <amp-story>.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function add_poster_images( $content ) {
+		$poster_images = Media::get_story_meta_images( $this->post );
+
+		unset( $poster_images['poster-portrait'] ); // Already exists.
+
+		foreach ( $poster_images as $attr => $url ) {
+			$attr_markup = sprintf( '%1$s-src="%2$s"', $attr, $url );
+			$content     = str_replace( 'poster-portrait-src=', $attr_markup . ' poster-portrait-src=', $content );
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Replaces the body start tag to fire a custom action.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function replace_body_start_tag( $content ) {
+		ob_start();
+
+		/**
+		 * Prints scripts or data after opening the <body>.
+		 */
+		do_action( 'web_stories_body_open' );
+
+		$output = (string) ob_get_clean();
+
+		return str_replace( '<body>', '<body>' . $output, $content );
+	}
+
+	/**
+	 * Replaces the body end tag to fire a custom action.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function replace_body_end_tag( $content ) {
+		ob_start();
+
+		/**
+		 * Prints scripts or data before closing the </body>.
+		 */
+		do_action( 'web_stories_footer' );
+
+		$output = (string) ob_get_clean();
+
+		return str_replace( '</body>', '</body>' . $output, $content );
```

Comment:
doesnt this be before body tag?

Reply:
Right! Good catch!

Reply:
Fixed in d9d79df90561c0f5e1aa1e761cef14820a914182